### ISSUE TITLE
[http_request] ESP32 reboots after few http requests 

### ISSUE THREAD

**Issue by XavierBerger** [esphome/issues#5951]
### The problem

TLDR: My code is working on 2024.5.5 and is crashing on 2024.6.1 after about  1.5min - http_request suspected to be the root cause.

---

I have a project using `http_request` and parsing `json` : 
  - [Source code available here](https://github.com/XavierBerger/Solar-Router-for-ESPHome/blob/main/esp32-standalone.yaml).
  - [http_request configuration available here](https://github.com/XavierBerger/Solar-Router-for-ESPHome/blob/main/solar_router/power_meter_fronius.yaml) - See YAML bellow

The code I used in my yaml to parse the json is coming from the official documentation.

As the size of json I read is about 1600 bytes, I did update `max_response_buffer_size`.  
The documentation of `max_response_buffer_size` let think that integer value is in kB : "`Default 1kB`"  
It looks like, in reality, the unit is bytes.  
I then set `max_response_buffer_size` to 4096 (I tried with 2048, issue is the same).

When I activate the http requests, after about a minute and half, my ESP is no more responding : No logs (even in debug), sensors unavailable in Web interface.  
After investigation, I understood that the ESP32 has rebooted.

I'm suspecting `http_request` because, if I do not activate the http requests, my ESP is responding (logs and sensors are updates). As soon as I start the http request, my ESP stop responding after about a minute and half.

The latest log I can see is always coming from http_request as for example: `10:58:26	[D]	[http_request.arduino:119]	
Content-Length: 1598`

### Which version of ESPHome has the issue?

2024.6.1

### What type of installation are you using?

Home Assistant Add-on 

and

esphome from pip

### Which version of Home Assistant has the issue?

2024.6.4

### What platform are you using?

ESP32-Arduino

### Board

ESP32-dev board - Wroom-32U

### Component causing the issue

http_request

### Example YAML snippet

```yaml
# ----------------------------------------------------------------------------------------------------
# Sensor updated every second to give feedback in Home Assistant
# ----------------------------------------------------------------------------------------------------

sensor:
  # Sensor showing the actual power consumption
  - id: real_power
    platform: template
    name: "Real Power"
    device_class: "power"
    unit_of_measurement: "W"
    update_interval: 1s

# ----------------------------------------------------------------------------------------------------
# Use http request component
# ----------------------------------------------------------------------------------------------------

http_request:
  id: http_request_data
  useragent: esphome/device
  timeout: 10s
  verify_ssl: False

# ----------------------------------------------------------------------------------------------------
# Define scripts for power collection
# ----------------------------------------------------------------------------------------------------

script:
  # Fronius script gather power reports from inverter and update globals (real_power)
  # Information are provided as json. Power exchanged with the grid is names PowerReal_P_Sum
  # When this value is positive, energy is taken from the grid.
  # When this value is negative, energy is pushed to the grid.
  - id: power_meter_source
    mode: single
    then:
      - if:
          condition:
            wifi.connected:
          then:
            - http_request.get:
                url: http://${power_meter_ip_address}/solar_api/v1/GetMeterRealtimeData.cgi
                capture_response: true
                max_response_buffer_size: 4096
                on_response:
                  then:
                    - lambda: |-
                        json::parse_json(body, [](JsonObject root) -> bool {
                          id(real_power).publish_state(root["Body"]["Data"]["0"]["PowerReal_P_Sum"].as< float >());
                          return true;
                        });

time:
  - platform: sntp
    on_time:
      - seconds: /1
        then:
          - if:
              condition:
                - lambda: return id(power_meter_activated) != 0;
              then:
                - script.execute: power_meter_source
```

### Anything in the logs that might be useful for us?

```txt
[11:42:04][D][light:036]: 'dimmer_controller' Setting:
[11:42:04][D][light:051]:   Brightness: 100%
[11:42:04][D][light:036]: 'green_led' Setting:
[11:42:05][D][light:036]: 'yellow_led' Setting:
[11:42:05][D][http_request.arduino:119]: Content-Length: 1600
WARNING solarrouter @ 192.168.1.27: Connection error occurred: [Errno 104] Connection reset by peer
INFO Processing unexpected disconnect from ESPHome API for solarrouter @ 192.168.1.27
WARNING Disconnected from API
INFO Successfully connected to solarrouter @ 192.168.1.27 in 0.011s
INFO Successful handshake with solarrouter @ 192.168.1.27 in 0.100s
[11:42:30][D][light:036]: 'yellow_led' Setting:
[11:42:30][D][light:047]:   State: OFF
[11:42:30][D][light:036]: 'yellow_led' Setting:
[11:42:30][D][light:047]:   State: ON
[11:42:35][D][light:036]: 'yellow_led' Setting:
```

### Additional information

I recompile the [v1.1.0](https://github.com/XavierBerger/Solar-Router-for-ESPHome/tree/v1.1.0) of my project with ESPHome 2.5.5 (pip) and everything is working well.

Here is the diff between the version which is working and the version which is crashing:

```diff
diff --git a/solar_router/power_meter_fronius.yaml b/solar_router/power_meter_fronius.yaml
index af3aa4e..0c6fad8 100644
--- a/solar_router/power_meter_fronius.yaml
+++ b/solar_router/power_meter_fronius.yaml
@@ -19,6 +19,7 @@ http_request:
   id: http_request_data
   useragent: esphome/device
   timeout: 10s
+  verify_ssl: False
 
 # ----------------------------------------------------------------------------------------------------
 # Define scripts for power collection
@@ -38,14 +39,14 @@ script:
           then:
             - http_request.get:
                 url: http://${power_meter_ip_address}/solar_api/v1/GetMeterRealtimeData.cgi
-                headers:
-                  Content-Type: application/json
-                verify_ssl: False
+                capture_response: true
+                max_response_buffer_size: 4096
                 on_response:
                   then:
                     - lambda: |-
-                        json::parse_json(id(http_request_data).get_string(), [](JsonObject data) {
-                          id(real_power).publish_state(data["Body"]["Data"]["0"]["PowerReal_P_Sum"].as< float >());
+                        json::parse_json(body, [](JsonObject root) -> bool {
+                          id(real_power).publish_state(root["Body"]["Data"]["0"]["PowerReal_P_Sum"].as< float >());
+                          return true;
                         });
 
 time:
```

**Comment by XavierBerger** [comment]
When I run the same code on ESP8285, I have other logs that may help in troubleshooting.

I did install this program with Home Assistant Addon.

Configuration:

```yaml
# ----------------------------------------------------------------------------------------------------
# ESPHome configuration - This part depends on your hardware target
# ----------------------------------------------------------------------------------------------------

esphome:
  name: "esp8285"
  friendly_name: ESP8285

esp8266:
  board: esp01_1m

# Enable logging
logger:
  baud_rate: 0
  level: INFO
  logs:
    component: ERROR

# Enable Home Assistant API
api:
  encryption:
    key: !secret esp8285_api_encryption_key

# Enable over-the-air updates
ota:

# WiFi connection
wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

# Activate web interface
web_server:
  port: 80

# ----------------------------------------------------------------------------------------------------
# Customisation
# ----------------------------------------------------------------------------------------------------

substitutions:
  # Power meter source -----------------------------------------------------------
  # Define ip address of Fronius inverter
  power_meter_ip_address: "192.168.1.21"

  # LEDs -------------------------------------------------------------------------
  # Yellow LED is reflecting power meter
  yellow_led_pin: GPIO1

globals:
  - id: power_meter_activated
    type: int
    initial_value: "1"

packages:
  power_meter:
    url: https://github.com/XavierBerger/Solar-Router-for-ESPHome/
    file: solar_router/power_meter_fronius.yaml
    ref: v1.1.0

http_request:
  esp8266_disable_ssl_support: True
```

The logs:

```
RAM:   [====      ]  43.1% (used 35276 bytes from 81920 bytes)
Flash: [=====     ]  51.7% (used 528989 bytes from 1023984 bytes)
Building .pioenvs/esp8285/firmware.bin
esp8266_copy_factory_bin([".pioenvs/esp8285/firmware.bin"], [".pioenvs/esp8285/firmware.elf"])
esp8266_copy_ota_bin([".pioenvs/esp8285/firmware.bin"], [".pioenvs/esp8285/firmware.elf"])
======================== [SUCCESS] Took 347.88 seconds ========================
INFO Successfully compiled program.
INFO Connecting to 192.168.1.30
INFO Uploading /data/build/esp8285/.pioenvs/esp8285/firmware.bin (533136 bytes)
INFO Compressed to 373065 bytes
Uploading: [============================================================] 100% Done...


INFO Upload took 10.95 seconds, waiting for result...
INFO OTA successful
INFO Successfully uploaded program.
INFO Starting log output from 192.168.1.30 using esphome API
WARNING Can't connect to ESPHome API for esp8285 @ 192.168.1.30: Timeout while connecting to [AddrInfo(family=<AddressFamily.AF_INET: 2>, type=<SocketKind.SOCK_STREAM: 1>, proto=6, sockaddr=IPv4Sockaddr(address='192.168.1.30', port=6053))] (TimeoutAPIError)
INFO Trying to connect to esp8285 @ 192.168.1.30 in the background
INFO Successfully connected to esp8285 @ 192.168.1.30 in 0.046s
INFO Successful handshake with esp8285 @ 192.168.1.30 in 2.655s
[20:35:35][I][app:100]: ESPHome version 2024.6.1 compiled on Jun 22 2024, 20:31:26
[20:35:35][W][safe_mode:028]: Last reset occurred too quickly; safe mode will be invoked in 8 restarts
[20:35:36][W][http_request.arduino:058]: Using HTTP on Arduino version >= 3.1 is **very** slow. Consider setting framework version to 3.0.2 in your YAML, or use HTTPS
[20:35:37][W][http_request.arduino:058]: Using HTTP on Arduino version >= 3.1 is **very** slow. Consider setting framework version to 3.0.2 in your YAML, or use HTTPS
[20:35:38][W][http_request.arduino:058]: Using HTTP on Arduino version >= 3.1 is **very** slow. Consider setting framework version to 3.0.2 in your YAML, or use HTTPS
[20:35:38][E][json:083]: Could not allocate memory for JSON document! Requested 1952 bytes, free heap: 1952
[20:35:39][W][http_request.arduino:058]: Using HTTP on Arduino version >= 3.1 is **very** slow. Consider setting framework version to 3.0.2 in your YAML, or use HTTPS
WARNING esp8285 @ 192.168.1.30: Connection error occurred: [Errno 104] Connection reset by peer
INFO Processing unexpected disconnect from ESPHome API for esp8285 @ 192.168.1.30
WARNING Disconnected from API
INFO Successfully connected to esp8285 @ 192.168.1.30 in 0.021s
INFO Successful handshake with esp8285 @ 192.168.1.30 in 2.727s
[20:36:26][W][http_request.arduino:058]: Using HTTP on Arduino version >= 3.1 is **very** slow. Consider setting framework version to 3.0.2 in your YAML, or use HTTPS
[20:36:27][W][http_request.arduino:058]: Using HTTP on Arduino version >= 3.1 is **very** slow. Consider setting framework version to 3.0.2 in your YAML, or use HTTPS
[20:36:28][W][http_request.arduino:058]: Using HTTP on Arduino version >= 3.1 is **very** slow. Consider setting framework version to 3.0.2 in your YAML, or use HTTPS
[20:36:29][W][http_request.arduino:058]: Using HTTP on Arduino version >= 3.1 is **very** slow. Consider setting framework version to 3.0.2 in your YAML, or use HTTPS
WARNING esp8285 @ 192.168.1.30: Connection error occurred: [Errno 104] Connection reset by peer
INFO Processing unexpected disconnect from ESPHome API for esp8285 @ 192.168.1.30
WARNING Disconnected from API
INFO Successfully connected to esp8285 @ 192.168.1.30 in 0.008s
WARNING Can't connect to ESPHome API for esp8285 @ 192.168.1.30: Handshake timed out after 30.0s (TimeoutAPIError)
INFO Trying to connect to esp8285 @ 192.168.1.30 in the background
WARNING esp8285 @ 192.168.1.30: Connection error occurred: esp8285 @ 192.168.1.30: Connection lost
INFO Successfully connected to esp8285 @ 192.168.1.30 in 0.006s
INFO Successful handshake with esp8285 @ 192.168.1.30 in 2.446s
[20:37:41][W][http_request.arduino:058]: Using HTTP on Arduino version >= 3.1 is **very** slow. Consider setting framework version to 3.0.2 in your YAML, or use HTTPS
```

**Comment by XavierBerger** [comment]
Issue is fixed for ESP32-dev board - Wroom-32U board.

On ESP8285 the warning about http_request.arduino velocity is continuously poping but has no impact on overall performances.

Issue is closed.
